plugins/inputs/mikrotik/mikrotik.go

```diff
@@ -0,0 +1,227 @@
+
+	return h.getClient()
+}
+
+func (h *Mikrotik) Gather(acc telegraf.Accumulator) error {
+	add, done, wait := getWaitGroup()
+	for _, u := range h.url {
+		add()
+		go func(url mikrotikEndpoint) {
+			defer done()
+
+			if err := h.gatherURL(url, acc); err != nil {
+				acc.AddError(wrapError("Gather", err))
+			}
+		}(u)
+	}
+	wait()
+
+	return nil
+}
+
+func (h *Mikrotik) getClient() (err error) {
+	tlsCfg, err := h.ClientConfig.TLSConfig()
+	if err != nil {
+		return wrapError("getClient", err)
+	}
+
+	h.client = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+		},
+		Timeout: time.Duration(h.ResponseTimeout),
+	}
+
+	return nil
+}
+
+func (h *Mikrotik) getSystemTags() error {
+	h.tags = make(map[string]string)
+	for _, tagURL := range h.systemTagsURL {
+		request, err := http.NewRequest("GET", tagURL, nil)
+		if err != nil {
+			return wrapError("getSystemTags", err)
+		}
+
+		err = h.setRequestAuth(request)
+		if err != nil {
+			return wrapError("getSystemTags", err)
+		}
+
+		binaryData, err := h.queryData(request)
+		if err != nil {
+			return wrapError("getSystemTags", err)
+		}
+
+		err = json.Unmarshal(binaryData, &h.tags)
+		if err != nil {
+			return wrapError("getSystemTags", err)
+		}
+	}
+	return nil
+}
+
+func (h *Mikrotik) gatherURL(endpoint mikrotikEndpoint, acc telegraf.Accumulator) error {
+	request, err := http.NewRequest("GET", endpoint.url, nil)
+	if err != nil {
+		return wrapError("gatherURL", err)
+	}
+
+	err = h.setRequestAuth(request)
+	if err != nil {
+		return wrapError("gatherURL", err)
+	}
+
+	binaryData, err := h.queryData(request)
+	if err != nil {
+		return wrapError("gatherURL", err)
+	}
+
+	timestamp := time.Now()
+
+	result, err := binToCommon(binaryData)
+	if err != nil {
+		return wrapError("gatherURL", err)
+	}
+
+	parsedData, err := parse(result)
+	if err != nil {
+		return wrapError("gatherURL", err)
+	}
+
+	for _, point := range parsedData {
+		point.Tags["source-module"] = endpoint.name
+		for n, v := range h.tags {
+			point.Tags[n] = v
+		}
+		acc.AddFields("mikrotik", point.Fields, point.Tags, timestamp)
+	}
+
+	return nil
+}
+
+func (h *Mikrotik) setRequestAuth(request *http.Request) error {
+	username := h.Username
+
+	password, err := h.Password.Get()
+	if err != nil {
+		return wrapError("setRequestAuth", err)
+	}
+	defer password.Destroy()
+
+	request.SetBasicAuth(username, password.String())
+
+	return nil
+}
+
+func (h *Mikrotik) queryData(request *http.Request) (data []byte, err error) {
+	resp, err := h.client.Do(request)
+	if err != nil {
+		return data, wrapError("queryData", err)
+	}
+
+	defer resp.Body.Close()
+	defer h.client.CloseIdleConnections()
+
+	responseHasSuccessCode := false
+
+	if resp.StatusCode == http.StatusOK {
+		responseHasSuccessCode = true
+	}
+
+	if !responseHasSuccessCode {
+		return data, wrapError("queryData", fmt.Errorf("received status code %d (%s), expected 200",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode)))
+	}
```

Comment:
Isn't this overly complex and corresponds to
```suggestion
	if resp.StatusCode != http.StatusOK {
		return data,  fmt.Errorf("server responded: %s (%d)", http.StatusText(resp.StatusCode)), resp.StatusCode)
	}
```